\chapter{Masai}
\label{sec:masai}

This chapter presents the engineering and evaluation of an efficient all-mapper for Illumina reads.
When I started this project, in October 2011, the fastest all-mappers (mrFast and RazerS~3) were two order of magnitude slower than popular best-mappers (Bowtie and BWA).
On the one hand, those all-mappers employed filtration based on exact seeds, which is efficient on short reference genomes but becomes too weak on mammal genomes; clearly, a stronger filtration method would had been beneficial.
On the other hand, those best-mappers are based on heuristic backtracking, which is inadequate to map longer Illumina reads.

After a thorough literature review, I came out with a novel read mapping method combining seed-based filtration with backtracking, published in the peer-reviewed journal \emph{Nucleic Acids Research} \citep{Siragusa2013}.
My method is packaged in a \CC tool nicknamed \emph{Masai}, which stands for \emph{m}ultiple backtracking of \emph{a}pproximate seeds on a \emph{s}uffix \emph{a}rray \emph{i}ndex.
Masai is part of the SeqAn library, it is distributed under the BSD license and can be downloaded from \url{http://www.seqan.de/projects/masai}.

In the engineering section, I explain Masai's indexing, filtration and verification methods for all-mapping.
In particular, the filtration method is based on approximate seeds: by employing approximate seeds instead of exact seeds, the tool obtains stronger, non-heuristic and quasi full-sensitive filtration for mammal reference genomes.
Masai find approximate seeds by backtracking the index of the reference genome.
The tool speeds up the backtracking phase by searching all seeds simultaneously, with the help of an additional index and the multiple backtracking algorithm.
Lastly, Masai implements also a quicker best-mapping method, though without mapping qualities.

In the evaluation section, I extensively compare Masai with popular read mappers, both on simulated and real datasets.
Compared to the all-mappers mrFast and RazerS~3, Masai is an order of magnitude faster and has comparable sensitivity.
In addition, Masai in best-mapping is 2--4 times faster and more accurate than Bowtie\,2 \citep{Langmead2012} and BWA \citep{Li2009}.
Finally, I discuss the limitations of Masai that led me to engineer yet another read aligner.

% -----------------------------------------------------------------------------

\section{Engineering}

I first give an outline of the read mapping method implemented in Masai.
Then, I explain each step in details, motivating relevant engineering choices that led me to the final implementation.

Masai requires an index capable of simulating a top-down traversal of the suffix trie of the reference genome.
The tool gives to users the possibility to choose among various indices (see section~\ref{masai:engineering:index}).
Similarly to all read mappers relying on an index of the reference genome, the tool indexes the reference genome only once, stores it on disk and reuses it for all subsequent read mapping jobs.

At mapping time, Masai requires two parameters to be provided: a maximum number of errors per read and a minimum seed length.
Default parameters work well for actual Illumina reads, otherwise the user has to adequately parametrize the tool for optimal performance.
Nonetheless, independently of the chosen parameterization, filtration is guaranteed to be quasi full-sensitive (see section~\ref{masai:engineering:seeding}).

Masai partitions all reads (and their reverse complements) into non-overlapping seeds and indexes them in a conceptual \emph{trie}.
Using the \emph{multiple backtracking} algorithm explained in section~\ref{sec:index:algo:multimismatch}, the tool backtracks simultaneously all indexed seeds in the suffix trie of the reference genome.
The program verifies all candidate locations, reported by the multiple backtracking algorithm, performing seed extension with a banded version of \emph{Myers bit-vector algorithm} \citep{Myers1999} (details in section~\ref{masai:engineering:extension}).

\subsection{Filtration}
\label{masai:engineering:seeding}

%I now consider formally the read mapping problem.
%Given a reference genome $g$, a set of reads $\mathcal{R}$ and an absolute number of errors $k$ consisting of indels and mismatches, for each read $r \in \mathcal{R}$ find all mapping locations where $r$ approximately occurs in $g$ within $k$ errors.

My original intent was to improve the speed of the all-mapper RazerS \citep{Weese2009} while preserving full-sensitivity under the edit distance.
RazerS was based on a $q$-gram filter;
I was aware that gapped $q$-grams could have brought a huge speedup, but I could not see any straightforward generalization of gapped $q$-grams to the edit distance.
At the same time, I experienced that weaker but quicker filtration using exact seeds is more advantageous than filtration using $q$-grams.
Indeed, a typical Illumina read mapping setup requires only moderate error rates, in the range of 4--6~\%.
For instance, RazerS~3 \citep{Weese2012} went back to filtration with exact seeds (similarly to mrFast).
Nonetheless, I wanted to improve filtration specificity of exact seeds, as the runtime of RazerS~3 on mammal genomes became dominated by verifications.
I knew that to improve filtration specificity I had to increase the seed length.

While reviewing past literature in the field of approximate string matching, I rediscovered the works of \cite{Myers1994} and \cite{Navarro2000} on approximate seeds, providing stronger filtration than exact seeds while preserving full-sensitivity under the edit distance.
Their idea is to partition the pattern into fewer non-overlapping seeds, which obviously can be longer than exact seeds but have to be searched approximately.
First, I slightly improved the filtration lemma of \citep{Navarro2000} to use approximate seeds with variable thresholds (see section~\ref{sec:seeds-apx}).
Then, as discussed in section~\ref{sec:seeds-apx}, I chose to parameterize Masai's filter by the seed length rather than by the number of seeds.
Indeed, the minimum seed length provides a direct estimate of the expected number of candidate locations to verify and thus of filtration specificity.
The resulting filter is thus flexible: by increasing the seed length, filtration becomes more specific at the expense of a higher filtration time.
In practice, the optimal seed length depends on the reference genome as well as on read lengths and the absolute number of errors.
In section \ref{masai:evaluation:filter}, I experimentally evaluate filtration schemes with exact and approximate seeds.
When mapping current Illumina reads on short to medium length genomes, exact seeds are still more efficient than approximate seeds.
Conversely, on larger genomes (\eg mammalian genomes) 1-approximate seeds outperform exact seeds by an order of magnitude.

Following the work of \cite{Navarro2000}, I decided to find approximate seeds by backtracking the suffix trie of the reference genome.
In section~\ref{masai:engineering:index}, I recall the engineering work I did to implement efficient backtracking of approximate seeds.
In order to achieve faster filtration, I opted to find approximate seeds only under the Hamming distance.
For this reason, when resorting to approximate seeds, Masai does not attain strict full-sensitivity under the edit distance.
Nonetheless, such implementation choice sacrifices only 0.1\% sensitivity (see section~\ref{masai:evaluation}).

\subsubsection{Best-mapping}

Masai is a tool primarily designed to perform all-mapping rather than best-mapping.
In best-mapping, Masai simply reports the first optimal mapping location encountered per read.
Clearly, this policy makes sense if the edit distance is effective at identifying the original mapping locations.
The evaluation of section~\ref{masai:evaluation:rabema} shows that Masai is competitive in best-mapping with tools using more complex scoring schemes.
Nonetheless, Masai's best-mapping method is ad-hoc and limited.
In best-mapping, the tool does not compute mapping qualities nor supports the paired-end and mate-pair protocols.
The reader is thus referred to the next chapter (section~\ref{sec:yara:eng}) for the complete description of an efficient best-mapping method that, in standard scenarios, is an order of magnitude faster than all-mapping.

\subsection{Indexing}
\label{masai:engineering:index}

Initially, the SeqAn library provided me with only two indices capable of simulating a top-down suffix trie traversal: the enhanced suffix array (ESA) \citep{Abouelhoda2004} and the lazy suffix tree (LST) \citep{Giegerich1999}.
To improve the efficiency of Masai, I implemented a generic top-down traversal for some additional indices, namely the suffix array (SA) \citep{Manber1990}, the $q$-gram index, and various specializations of the full-text minute index (FM-index) \citep{Ferragina2001}.
Below I discuss the performance of these indices within Masai, while I refer the reader to chapter~\ref{chr:index} for their extensive explanation.

\subsubsection{Indexing the reference genome}

I initially chose the ESA over the LST because of better construction times.
Indeed, the ESA provides a linear time construction algorithm (an adaptation of the DC7 algorithm \citep{Dementiev2008} to multiple sequences \citep{Weese2013} for the generalized SA, followed by the algorithms proposed in \citep{Kasai2001,Abouelhoda2004}), while the LST construction algorithm takes quadratic time (using the radix sort based \emph{wotd}-algorithm \citep{Giegerich1999}).
The construction of the ESA of the \emph{H.~sapiens} reference genome (GRCh37) takes about 1.5~hours and the index consumes 39~GB of memory.% (see table~\ref{tab:index:construction}).
Apart from that, both ESA and LST implementations require $13$ bytes per base pair and exhibit comparable query speed.

At this point, Masai required high-end hardware to process large reference genomes.
Therefore, thinking about a space-time trade-off, I designed a generic suffix trie top-down traversal for the SA (see section~\ref{sec:index:sa}).
The SA consumes only $5$ bytes per base pair but is theoretically slower than the ESA, as it adds a logarithmic factor to query times.
However, with surprise, I found out that within Masai the SA had equal or better performance than the ESA (see table~\ref{tab:masai:indices}).
Ultimately, this change brought down the memory footprint of the index from 39~GB to 15~GB but preserved query speed.

I tried to further improve query speed by removing the logarithmic factor introduced by the SA.
Therefore, to cut the most expensive binary searches, I put a $q$-gram index on top of the SA and extended my generic suffix trie top-down traversal accordingly (see section~\ref{sec:index:qgram}).
Yet, the $q$-gram index did not bring significant speedup to the application;
indeed, the lookup table turned out to be useful when searching patterns one by one, but not when coupled with the multiple backtracking algorithm.

Finally, I explored additional space-time trade-offs.
Starting from the implementation of \citep{Singer2012}, I realized a generalized FM-index based on a wavelet tree~\citep{Grossi2003}.
This initial FM-index consumed about $1.5$ bytes per base pair with a SA sampling of 10~\%.
Thus the memory footprint of the index went down to 4.5~GB, but Masai became almost twice as slow (see table~\ref{tab:masai:indices}).

To sum up, I preferred the SA as it provides a good compromise between query speed and memory consumption.
Nevertheless, Masai leaves to the user the possibility of choosing among the aforementioned data structures.
Table \ref{tab:masai:indices} summarizes the runtime of the program with various indices.

\subsubsection{Indexing the reads}

In order to improve index query speed, I designed and implemented the algorithms presented in sections \ref{sec:index:algo:multiexact} and \ref{sec:index:algo:multimismatch}.
These algorithms search simultaneously many exact or approximate seeds, achieving a speedup of 2--5 times over their conventional counterparts.
As this multiple string matching algorithm requires a trie of the seeds, I also engineered an efficient trie implementation.
A short explanation can be found in section \ref{sec:index:trie}

Building the SA via quicksort turned out to be faster than building the LST via radix sort but, within Masai, the more involved LST data structure payed off in terms of query time (see section~\ref{sec:index:algo:traversal}).
Indeed, the LST stores all trie nodes and thus provides node traversal in constant time, while the SA explicitly stores only the leaves and then derives internal nodes via binary search.
As the memory footprint of the trie is negligible within this application, I chose the LST to perform multiple backtracking of approximate seeds.

When performing multiple backtracking of exact seeds, the LST construction time dominates the overall filtration time (see section~\ref{sec:index:algo:multiexact}).
Therefore, I decided to resort to the \emph{$q$-gram index} to emulate a trie in this case:
I build a partial $q$-gram index efficiently and in linear time by bucket sort, again considering only the first suffix of each seed in the collection.
Such index represents a trie truncated at depth $q$ (which I fixed to 12 in Masai).
Truncation is only a minor concern: at depth $q$ the search continues separately on each active node using the conventional binary search algorithms (see sections \ref{sec:index:algo:exact} and \ref{sec:index:algo:kmismatches}).
%Given the sparseness of the seeds index, for exact search it pays off to adopt a trie based on a $q$-gram index rather than a LST-based trie.

\subsection{Verification}
\label{masai:engineering:extension}

To verify candidate locations reported by the filtration algorithm, Masai employs a banded version of Myers bit-vector algorithm \citep{Myers1999} implemented for the tool RazerS~3 by \citeauthor{Weese2012} and available in the SeqAn library.
This banded version computes only a diagonal band of the DP matrix and is thus faster than the original algorithm by Myers.
However, Masai uses this algorithm in a different way than RazerS~3.
While RazerS~3 performs one semi-global alignment to verify one parallelogram surrounding any seed occurrence, Masai performs two global alignments on both ends of any seed occurrence.
Given a seed occurring with $e$ errors, Masai first performs seed extension on the left side within an error threshold of $k - e$ errors.
Only if the seed extension on the left side succeeds, the tool performs a seed extension on the right side within the remaining error threshold.
Moreover, the tool first computes the \emph{longest common prefix} on each side of the seed and lets the global alignment algorithm start from the first mismatching positions.
This approach is up to two times faster than the one implemented by RazerS~3 (data not shown).

% -----------------------------------------------------------------------------

\section{Evaluation}
\label{masai:evaluation}

In order to evaluate Masai, I propose three experiments:
\begin{inparaenum}[(i)]
\item the Rabema benchmark on simulated data,
\item variant detection on simulated data, and 
\item performance on real data.
\end{inparaenum}
This evaluation focuses on the capability of the mappers to retrieve the location of a single read without the help of its paired-end, which can of course disambiguate some mapping locations.
As references, I use whole genomes of \emph{E.~coli} (NCBI NC\_000913.2), \emph{C.~elegans} (WormBase WS195), \emph{D.~melanogaster} (FlyBase release 5.42), and \emph{H.~sapiens} (GRCh37.p2).

I compare Masai in all-mapping with RazerS\,3, Hobbes, mrFAST and SHRiMP\,2, while in best-mapping with Bowtie\,2, BWA and Soap\,2.
Masai, RazerS\,3, Hobbes and mrFAST use edit distance, while Bowtie\,2, BWA, Soap\,2 and SHRiMP\,2 rely on scoring schemes taking into account base quality values.
When relevant, I configured some read mappers with the appropriate absolute number of errors (Masai, mrFAST, Hobbes, Soap\,2) or error rate (RazerS\,3).
In section \ref{sup:masai:param}, I give the exact parameterization of the read mappers considered in this evaluation.

\subsection{Rabema benchmark on simulated data}
\label{masai:evaluation:rabema}

I first consider the Rabema benchmark~\citep{Holtgrewe2011} (v1.1) for a thorough evaluation and comparison of read mapping sensitivity.
I consider the benchmark categories \emph{all} and \emph{best}, in addition to \emph{precision} and \emph{recall}.
In the categories all and best, a read mapper has to find for each read respectively all or one of the best edit distance mapping locations.
The categories precision and recall require a read mapper to find the \emph{original} location of each simulated read, which is a measure independent of the used scoring model, \eg edit distance or quality based.
A simulated read is mapped \emph{correctly} if the mapper reports its original location, 
and it is mapped \emph{uniquely} if the mapper reports only one location.
Rabema defines \emph{recall} to be the fraction of reads which were correctly mapped and \emph{precision} the fraction of uniquely mapped reads that were mapped correctly.

Similarly to \citep{Langmead2012}, I used the read simulator Mason \citep{Holtgrewe2010} with default profile settings to simulate, from each whole genome, 100\,k reads of length 100\,bp having sequencing errors distributed like in a typical Illumina run.
I performed the benchmark for an error rate of 5\,\%, which corresponds to edit distance 5 for reads of length 100\,bp. Therefore, I built a Rabema gold standard for each dataset by running RazerS~3 in full-sensitive mode up to edit distance 5. I further classified mapping locations in each category by their edit distance.

For a more fair and thorough comparison, I also consider BWA and Bowtie\,2 in all-mapping (Soap\,2 cannot be configured accordingly).
To this extent, I parametrized these tools to be highly sensitive and output all found mapping locations.
Since BWA and Bowtie\,2 were not designed to be used in this way, they spent much more time than proper all-mappers, \ie up to 3~hours in a run compared to several minutes.
However, the aim of this experiment is to investigate read mapping sensitivity, therefore I do not report any running times.
Table~\ref{tab:masai:rabema} shows the results on \emph{H.~sapiens}.

\subsubsection{All-mapping}
As expected, RazerS\,3 shows full-sensitivity.
In contrast, mrFAST looses a minimal percentage of mapping locations.
Overall, Masai does not loose more than 0.1\,\% of all mapping locations.
In particular, Masai is full-sensitive for low-error locations and looses only a small percentage of high-error locations, \ie its loss is limited to 0.1\,\% and 1.4\,\% of mapping locations at edit distance 4 and 5.

Conversely, BWA and Bowtie\,2 miss 35\,\% and 45\,\% of all mapping locations at edit distance 5 and their recall values as all-mappers do not substantially increase.
Likewise, SHRiMP\,2 is not able to enumerate all mapping locations, although its recall values are good.
Again, Hobbes has the worst performance.

As mentioned in section \ref{masai:engineering:seeding}, Masai is not full-sensitive whenever approximate seeds are used, \eg on \emph{H.~sapiens}. Indeed, Masai loses 0.1\,\% overall sensitivity in respect to RazerS\,3.
%Conversely, it attains full-sensitivity whenever exact seeds are used, \eg on E.~coli, C.~elegans and D.~melanogaster (Supplementary Data).
In general, RazerS\,3 should be used when full-sensitivity is required, \ie in read mapping benchmarks.
However, these results show that Masai can replace RazerS\,3 or mrFAST in practical all-mapping setups.

\subsubsection{Best-mapping}
Masai shows the best recall values, not loosing more than 2.3\,\% recall on edit distance 5.
Conversely, the recall values of BWA and Bowtie\,2 drop significantly with increasing edit distance and loose up to 15.4\,\% and 11.5\,\% on edit distance 5.
As expected, Soap\,2 turns out to be inadequate for mapping reads of length 100\,bp at this error rates.
Precision values have less variance than recall values. Masai shows the best precision values with 97.8\,\%, followed by Soap\,2 with 97.7\,\%, and BWA with 97.5\,\%. Interestingly, Bowtie\,2 shows the worst precision values, loosing up to 5.6\,\% on edit distance 5.

\begin{table*}[t]
  \caption[Masai results in the Rabema benchmark]
  {
  \label{tab:masai:rabema}
    Rabema benchmark results on $100\,\text{k}\times 100\,\text{bp}$ Illumina-like reads.
    Rabema scores are given in percent (average fraction of edit distance locations reported per read).
    Large numbers show total scores in each Rabema category and small numbers show the category scores separately for reads with $\bigl(\begin{smallmatrix}\mbox{\tiny 0}&\mbox{\tiny 1}&\mbox{\tiny 2}\\\mbox{\tiny 3}&\mbox{\tiny 4}&\mbox{\tiny 5}\end{smallmatrix}\bigr)$ errors.
    }
  \vspace{-3mm}
  \center
  \sffamily
  \footnotesize
  \resizebox{0.95\textwidth}{!}
  {
	\renewcommand{\tabcolsep}{0.8ex}
	\input{tables/table_masai_rabema}
  }
\end{table*}

\subsection{Variant detection on simulated data}
\label{masai:evaluation:variant}

The second experiment analyzes the theoretical performance of Masai and other read mappers in variant detection pipelines.
Similarly to \citep{David2011}, this experiment considers simulated reads containing sequencing errors, SNPs and indels.
Each simulated read has an edit distance of at most 5 to its genomic origin, and it is grouped according to the number of contained SNPs and indels, where class $(s,i)$ consists of all reads with $s$ SNPs and $i$ indels.
The experiment considers a read to be mapped \emph{correctly} if a mapping location is reported within 10\,bp of its genomic origin;
it considers a read to map \emph{uniquely} if only one location is reported by the mapper.
For each class, the experiment defines \emph{recall} to be the fraction of reads which were correctly mapped and \emph{precision} the fraction of uniquely mapped reads that were mapped correctly.

I simulated 5 million Illumina-like reads of length $100$\,bp from the whole human genome using Mason.
I mapped the reads with each tool and measured its sensitivity in each class.
Table~\ref{tab:masai:variant} shows the results.

\subsubsection{All-mapping}

Looking at all-mapping results, Masai shows 100\,\% precision and recall in all classes, except for classes (2,0) and (1,1) where it looses only 0.1\,\% and 0.7\,\% recall.
Masai is therefore roughly comparable to the full-sensitive read mappers RazerS\,3 and mrFAST.
SHRiMP\,2 shows 100\,\% precision in all classes but looses between 0.3\,\% and 0.8\,\% recall in each class.
Hobbes has the lowest performance among all-mappers: it appears to have problems with indels, indeed it looses 9.5\,\% recall in class (0,3).

\subsubsection{Best-mapping}

Masai shows the highest precision and recall in all best-mapping classes.
In particular, Masai does not loose more than 3.2\,\% recall in class (4,0), whether Bowtie\,2 and BWA loose respectively 17.5\,\% and 14.9\,\% and Soap\,2 is not able to map any read.
The recall values of Bowtie\,2, BWA and Soap\,2 are negatively correlated with the amount of genomic variation.
For instance, in the Rabema benchmark, Bowtie\,2 looses respectively 7.2\,\% and 11.5\,\% of mapping locations at distance 4 and 5, but in class (4,0) of this experiment it looses 17.5\,\% recall.
A similar trend is observable for BWA and Soap\,2.
The low performance of Soap\,2 is also due to its limitation to at most 2 mismatches and no support for indels.

\begin{table*}[tH!]
  \caption[Masai variant detection results]
  {
  \label{tab:masai:variant}
    Variant detection results on $5\,\text{M}\times 100\,\text{bp}$ Illumina-like reads.
    The table shows percentages of found origins (recall) and fraction of unique reads mapped to their origin (precision) classed by reads with $s$ SNPs and $i$ indels $(s,i)$.
  }
  \vspace{-3mm}
  \center
  \sffamily
  \small
  \resizebox{\textwidth}{!}
  {
	\renewcommand{\tabcolsep}{0.8ex}
	\input{tables/table_masai_variant}
  }
\end{table*}

\subsection{Performance on real data}
\label{masai:evaluation:performance}

The last experiment focuses on comparing read mappers performance on real data.
I mapped the first $10\,\text{M}\times 100\,\text{bp}$ reads from an Illumina lane of \emph{E.~coli} (ERR022075, Genome Analyzer IIx), \emph{D.~melanogaster} (SRR497711, HiSeq 2000), \emph{C.~elegans} (SRR065390, Genome Analyzer II), and \emph{H.~sapiens} (ERR012100, Genome Analyzer II).
Whenever possible, I configured the tools to map the reads within edit distance 5.
I measured mapping times on a cluster of nodes with 72\,GB RAM and 2 Intel Xeon X5650 processors running Linux~3.2.0.
For an accurate running time comparison, I ran the tools using a single thread and used local disks for I/O.
I measured running times and peak memory consumptions.

I cannot measure precision and recall values as real reads have unknown origins.
%Therefore, for this evaluation, I adopt the commonly used measure of percentage of \emph{mapped reads}, \ie the fraction of reads for which the read mapper reports a mapping location.
%However, as some mappers report mapping locations without constraints on the number of errors, 
Therefore, for this evaluation, I count also the Rabema \emph{best} locations.
As explained in section \ref{masai:evaluation:rabema}, the Rabema best category assigns a point for a read if the mapper reports at least one mapping location with the optimal (\ie best) number of errors;
final Rabema best scores are normalized by the number of reads.
All results are shown in table~\ref{tab:masai:runtime}.

\subsubsection{All-mapping}
On the \emph{H.~sapiens} dataset, Masai is 11.9 times faster than RazerS\,3, 14.6 times faster than mrFAST, and 7.6 times faster than Hobbes.
The current version of Hobbes constantly crashes and maps only half of the reads.
SHRiMP\,2 is not able to map the \emph{H.~sapiens} dataset within 4 days.
On the \emph{C.~elegans} dataset, Masai is 2.0 times faster than RazerS\,3, 10.9 times faster than Hobbes, 6.3 times faster than mrFAST and 50.1 times faster than SHRiMP\,2.
Hobbes constantly crashes and maps less reads than all other mappers in this category.

%Likewise for Bowtie\,2, also SHRiMP\,2 does not impose a maximal error rate threshold and reports more mapped reads than Masai.
%Also SHRiMP\,2 Rabema best score is inferior to Masai.
%This could be due to the use of a different scoring scheme where two mismatches cost less than opening a gap.
%Anyway, this hypothesis does not explain why SHRiMP\,2 does not report some mapping locations at distance 0.

\subsubsection{Best-mapping}
On the \emph{H.~sapiens} dataset, Masai is 2.6 times faster than Bowtie\,2, 3.6 times faster than BWA but 2.1 times slower than Soap\,2.
On the \emph{C.~elegans} dataset, Masai is 7.7 times faster than Bowtie\,2, 8.2 times faster than BWA and 1.5 times faster than Soap\,2.
%On one end, 
Soap\,2 is not able to map a consistent fraction of reads because of its limitation to 2 mismatches.
%On the other end, Bowtie\,2 reports more mapped reads than Masai but, taking best scores into account, it reports less mapping locations than Masai.
%In fact, Bowtie\,2 uses a scoring scheme based on quality values and does not impose a maximal error rate threshold.
%Despite that, 
Bowtie\,2 misses respectively 22.0\,\% and 20.7\,\% of reads mappable at edit distance 5 on the \emph{C.~elegans} and \emph{H.~sapiens} datasets.

\begin{table*}[t]
  \caption[Masai performance on real data]{
    \label{tab:masai:runtime}
    Performance on real data using $10\,\text{M}\times 100\,\text{bp}$ Illumina reads.
	Rabema best column: in large are shown the percentage of reads mapped with the minimal number of errors (up to 5\%) and in small the percentage of reads that were mapped with $\bigl(\begin{smallmatrix}\mbox{\tiny 0}&\mbox{\tiny 1\%}&\mbox{\tiny 2\%}\\\mbox{\tiny 3\%}&\mbox{\tiny 4\%}&\mbox{\tiny 5\%}\end{smallmatrix}\bigr)$ errors.
%	Mapped reads column: in large are shown the percentage of mapped reads and in small the cumulative percentage of reads that were mapped with $\bigl(\begin{smallmatrix}\mbox{\tiny 0}&\mbox{\tiny 1\%}&\mbox{\tiny 2\%}\\\mbox{\tiny 3\%}&\mbox{\tiny 4\%}&\mbox{\tiny 5\%}\end{smallmatrix}\bigr)$ errors.
	Remarks: SHRiMP\,2 is not able to map the \emph{H.~sapiens} dataset within 4 days; Hobbes systematically crashes in all but the \emph{D.~melanogaster} dataset.
  }
	\vspace{-3mm}
	\center
	\sffamily
	\footnotesize
	\resizebox{0.95\textwidth}{!}
	{
		\renewcommand{\tabcolsep}{0.8ex}
		\input{tables/table_masai_runtime}
	}
\end{table*}


\subsection{Performance with different indices}
\label{masai:evaluation:index}

Table~\ref{tab:masai:indices} shows the runtime and memory consumption of Masai using different indices.
As discussed in section~\ref{masai:engineering:index}, Masai/SA is always faster than Masai/ESA, except for all-mapping on the \emph{H.~sapiens} dataset.
Masai/FM-index shows significantly lower memory consumption only on the \emph{H.~sapiens} dataset, but with a consistent runtime penalty.

\begin{table*}[t]
  \center
  \caption[Masai performance with different indices]{
    \label{tab:masai:indices}%
    Masai performance with different indices. The lowest runtimes are in bold.
  }
  {
  \sffamily
  \small
  \renewcommand{\tabcolsep}{0.8ex}
  \input{tables/table_masai_indices}
  }
  \vspace{2mm}
\end{table*}


\subsection{Filtration efficiency}
\label{masai:evaluation:filter}

This experiment assesses the contribution of approximate seeds and multiple backtracking on runtime results.
To this intent, I perform all-mapping with Masai on all dataset of section~\ref{masai:evaluation:performance}, using either exact or approximate seeds in combination with either single or multiple backtracking.
As this experiment focuses on filtration efficiency, I do not consider the time spent performing seed extensions and I/O, \ie loading the reference genome and its index, loading the reads, and writing the results.

Table~\ref{tab:masai:filtration} shows the results.
Column \emph{time} reports filtration times, \ie the time spent to index the seeds (in case of multiple backtracking) and to perform backtracking.
Column \emph{candid.} reports the number of candidate locations reported by the filter for which seed extension is subsequently performed.
In bold, I report the optimal filtering scheme used to parameterize Masai in the experiments of section \ref{masai:evaluation:performance}.

On \emph{E.~coli}, \emph{D.~melanogaster} and \emph{C.~elegans}, approximate seeds reduce the number of candidates respectively by 2.1 times, 9.9 times, and 4.3 times.
Nevertheless I still prefer exact seeds as filtration dominates the total runtime.
Multiple backtracking on exact seeds compared to single backtracking speeds up filtration by 2.9 times on \emph{E.~coli}, and 3.8 times on \emph{D.~melanogaster} and \emph{C.~elegans}.
Without the contribution of multiple backtracking, Masai would not be faster than RazerS\,3, the second fastest all-mapper.

Approximate seeds become effective on \emph{H.~sapiens}, where they reduce the number of candidates by 10.8 times. 
On \emph{H.~sapiens}, seed extensions largely dominate the total runtime, therefore I prefer approximate seeds.
Multiple backtracking on approximate seeds provides a speed-up of 3.2 times over single backtracking.
The combination of the two methods makes Masai an order of magnitude faster than any other all-mapper.

\begin{table*}[b]
  \center
  \caption[Masai filtration efficiency results]{
    \label{tab:masai:filtration}%
    Masai all-mapping filtration efficiency results.
    Filtering times include seeds indexing times.
    The best filtering schemes are in bold.
  }
  \resizebox{\textwidth}{!}
  {
  \sffamily
  \small
  \renewcommand{\tabcolsep}{0.8ex}
  \input{tables/table_masai_filters}
  }
  \vspace{2mm}
\end{table*}


\section{Discussion}

Masai consists of two important algorithmic methods: approximate seeds and multiple backtracking.
Approximate seeds are of paramount importance to obtain very specific, yet full-sensitive filtration; their adoption speeds up Masai by one order of magnitude.
Multiple backtracking further speeds up the filtration phase by 3--5 times on a (enhanced) suffix array index; this technique makes Masai twice as fast.
In addition, Masai implements a best-mapping method that finds one optimal mapping location and is an order of magnitude faster than all-mapping.

Is the edit distance adequate for best-mapping?
Both Rabema benchmark and variant detection results show that Masai has constantly better accuracy than other best-mappers relying on more complex scoring schemes.
In particular, the Rabema benchmark results show that Rabema best values are tightly bound to recall scores.
Hence, the edit distance is a pertinent and adequate scoring scheme for best-mapping.
Vice versa, best-mappers using scoring schemes based on quality values show a generalized and substantial loss of mapping accuracy.
This is likely due to the heuristics on which these tools rely.
To sum up, it is better to stick to edit distance and guarantee full-sensitivity rather than to adopt an involved scoring scheme and explore the alignment space heuristically, hence partially.

How many mapping locations do heuristic best-mappers miss?
By looking at precision and recall values on simulated data, or at Rabema best values on real data, it can be deduced that Bowtie\,2, BWA and Soap\,2 miss up to 20\,\% of reads mappable at 5\,\% error rate.
Yet, it is not evident how these results affects variant calling pipelines.

Summing up, Masai in all-mapping is an order of magnitude faster and thus a valid alternative to tools like RazerS~3 and mrFast.
Computational requirements of all-mapping are now close to those of best-mapping.
Indeed, Masai in all-mapping is only 4 times slower than BWA in best-mapping, despite reporting two orders of magnitude more mapping locations.
Masai in best-mapping is 2--4 times faster and more accurate than Bowtie\,2 \citep{Langmead2012} and BWA \citep{Li2009}.
The achieved speedup is huge when RazerS~3 is used for best-mapping: in this scenario, Masai is roughly 200 times faster!

Despite these good results, Masai is not being widely used.
This is mainly because the tool lacks some commonly requested features, including:
direct support of the paired-end and mate-pair protocols, computation of mapping qualities, parallelization via multi-threading, low memory footprint, and automatic parameterization.
Because of my initial inexperience and unclear or wrong design goals, I neglected these features while engineering Masai.
The next chapter introduces \emph{Yara}, a tool implementing these features.
